JP-3701: Blend models

braingram opened the pull request:
Refactor model blender to allow incremental model blending (and use it for resample).

Regression tests show only failures that exist on main:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1695/

EDIT: new regtests https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1698/

~Note that this does not resolve https://jira.stsci.edu/browse/JP-3732 as fixing the observation time blending causes several regression tests to fail. If preferred this PR could be updated to fix JP-3732 by replacing `bad_mintime` with `min` (and `bad_maxtime` with `max).~ EDIT: fixed in PR

~This PR also duplicates the existing behavior where any attribute with `blend_table: True` in the schema will be accessed during blending. This creates these attributes (if the schema defines a default) in models that don't otherwise have these attributes defined. One example is [meta.instrument.lamp_mode](https://github.com/spacetelescope/stdatamodels/blob/a49474ce4f5f63423ac20a82ce9dd053bf89d3cf/src/stdatamodels/jwst/datamodels/schemas/core.schema.yaml#L567). When blending models if they don't have `meta.instrument.lamp_mode` set, after blending they will have `meta.instrument.lamp_mode` set to the string "NONE".~ EDIT: fixed in PR

<!-- If this PR closes a JIRA ticket, make sure the title starts with the JIRA issue number,
for example JP-1234: <Fix a bug> -->
Resolves [JP-3701](https://jira.stsci.edu/browse/JP-3701)
Resolves [JP-3732](https://jira.stsci.edu/browse/JP-3732)

<!-- If this PR closes a GitHub issue, reference it here by its number -->
Closes #8685

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [ ] added entry in `CHANGES.rst` within the relevant release section
- [ ] updated or added relevant tests
- [ ] updated relevant documentation
- [ ] added relevant milestone
- [ ] added relevant label(s)
- [ ] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [ ] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

emolter commented on jwst/model_blender/tablebuilder.py:
```diff
@@ -0,0 +1,57 @@
+    }
+
+
+class TableBuilder:
```
can you add short docstrings to this class and its functions?

braingram replied:
Good idea! I added docstrings in: https://github.com/spacetelescope/jwst/pull/8759/commits/68d9f6b149b3c1d4b57a330a524c19d831b10799
(and added some for `ModelBlender`).

emolter commented on jwst/model_blender/schemautil.py:
```diff
@@ -0,0 +1,60 @@
+        will combine all 'meta.observation.time' attributes
+        using the 'mintime' rule.
+
+    schema_ignores: lsit
```
```suggestion
    schema_ignores: list
```

emolter commented on jwst/model_blender/blendmeta.py:
```diff
@@ -3,328 +3,16 @@
-    results = []
-    dm_schema.walk_schema(schema, build_rules_list, results)
-    return results
+def blendmodels(product, inputs, ignore=None):
```
Since this is now only a single function of five lines, would it perhaps be easier and enhance readability to delete this file and move the function elsewhere, e.g. into blender.py?

emolter replied:
Actually, is this used anywhere besides the test suite?  Could it be removed entirely?

braingram replied:
Thanks! This function is used in cube_build/ifu_cube.py. See:https://github.com/spacetelescope/jwst/pull/8759/files#diff-4eb3446a610f71b161f1fdbe2ff28934ea50381e54a060851e440882bb272ae1R2366

emolter replied:
huh, not sure why that didn't pop up when I clicked the button that's supposed to show me where it's used. thanks.

emolter commented on jwst/model_blender/rules.py:
```diff
@@ -0,0 +1,81 @@
+    if num_vals == 1:
+        return uniq_vals[0]
+    if num_vals > 1:
+        return "MULTIPLE"
```
Could a unit test for the Blender class be added that touches the `"MULTIPLE"` return and ensures that everything works properly when the list isn't identical?

braingram replied:
Good idea! I added one in: d75874c48

emolter commented on jwst/model_blender/rules.py:
```diff
@@ -0,0 +1,81 @@
+
+
+def bad_mintime(vals):
+    # this introduces floating point error for backwards compatibility
```
I think it's fine to make this change in the same PR and just say it resolves both tickets, but I suppose that's a procedural question best left to Melanie and Tyler

braingram replied:
@melanieclarke @tapastro 

Would you like me to fix https://jira.stsci.edu/browse/JP-3732 in this PR? It would require updating regtests.

If we're updating regtests I'd also like to consider fixing the missing attribute creation mentioned in the PR description. It seems like a bug to me that blending models would create attributes in the input models (even if the created values match the defaults).

melanieclarke replied:
Okay by me - better to fix the bug than go out of your way to replicate it!

melanieclarke replied:
Although I do appreciate that you were able to get a clean set of regression tests first to make sure everything else was working.  Thanks for that!

emolter commented on jwst/model_blender/rules.py:
```diff
@@ -0,0 +1,81 @@
+    'max': np.max,
+    'min': np.min,
+
+    # these intentionally introduce floating point errors
+    # for backwards compatibility but would be more accurate
+    # using the builtin min/max
```
```suggestion
```
probably only necessary to include this once

emolter reviewed (commented):
Thanks for the hard work Brett, this looks great to me.  The code is less verbose and also a lot easier to read than before.  I'm definitely not an expert in how the blender works or used to work, so most of these comments are just requests for additional docstrings and typo catches.

It does appear that the changes will ensure it's not necessary to load all models at once into memory, and my hacky workaround can be removed.  So I agree this resolves the ticket according to my original intention 🥳

braingram commented on docs/jwst/model_blender/main.rst:
```diff
@@ -27,86 +35,73 @@ This code works by
-JWST environment not be correct for the user's input data.  The user can simply
-edit the schema definition installed in their JWST environment to apply custom
-rules for blending the data being processed.
+If an input model is missing an attribute a ``nan`` will be stored in the
```
> Does "missing an attribute" mean there is an attribute in the schema that doesn't exist in the model?

Yeah, so for an attribute that has `blend_table: True` in the schema, if the datamodel does not have that attribute the table cell (for the row for that model) will contain a `nan`.

> Is the NaN stored only if none of the input models have the attribute, or if the first model doesn't have it, or something else?

The `nan` is only for the table cell for the row corresponding to the datamodel that is missing the attribute.

braingram commented on jwst/model_blender/blender.py:
```diff
@@ -1,254 +1,94 @@
+    def accumulate(self, model):
+        if self._first_header_meta is None:
+            # search the schema for other metadata to "blend" and to add to the table
+            attr_to_columns, attr_to_blend_rules, schema_ignores  = parse_schema(model.schema)
```
Blending of models with different types isn't supported.

I added another warning to the docs that all models must be the same type and a check in `ModelBlender` (and unit test for the check) in: d13895205848b163a3a1c09e7901325279d0fe23

braingram commented:
I fixed the 2 bugs found while working on this PR (the ones mentioned in the PR description).

This results in many regtest "failures":
https://plwishmaster.stsci.edu:8081/blue/organizations/jenkins/RT%2FJWST-Developers-Pull-Requests/detail/JWST-Developers-Pull-Requests/1698/tests

with "[stable-deps] test_miri_lrs_slit_spec3[default_wcs-x1d] – jwst.regtest.test_miri_lrs_slit_spec3" being a good example showing both bugs. With this PR the fitsdiff fails with:
```
     Headers contain differences:
       Headers have different number of cards:
        a: 256
        b: 257
       Extra keyword 'OPMODE' in b: 'NONE'
       Keyword TIME-OBS has different values:
          a> 23:07:34.995
           ?            ^
          b> 23:07:34.994999
           ?   
```
The `TIME-OBS` in the result (a) is more accurate than truth (b) and model blender no longer creates the default `OPMODE` (which doesn't exist in the input, blended models).

There are also many unrelated regtest failures in this run due removal of `R_DRZPAR`, what looks like a crds update and existing failures.

melanieclarke commented on jwst/model_blender/blender.py:
```diff
@@ -1,254 +1,152 @@
+        ----------
+
+        model: `jwst.datamodels.JwstDataModel`
+            The datatamodel to blend.
```
```suggestion
            The datamodel to blend.
```

melanieclarke commented on jwst/model_blender/rules.py:
```diff
@@ -0,0 +1,65 @@
+}
+
+
+class Blender:
```
Also either add docs to the RULE_FUNCTIONS, since the blender_rules docs refer to it, or else describe them long-form in blender_rules.rst.

braingram replied:
I added `RULE_FUNCTIONS` to the docs in 20db503f8 and added docs for `Blender` and `make_blender` in c8e0fefa06e007a6fe2e48e71a0ed8696d7abd77

melanieclarke commented on jwst/model_blender/rules.py:
```diff
@@ -0,0 +1,65 @@
+        return "MULTIPLE"
+
+
+def _isotime(time_str):
```
This function doesn't appear to be used - can we remove it?

braingram replied:
Thanks for the catch. I removed this in d08c42dc5 It was previously used by the `bad_min/maxtime` and could have been removed when those were removed.

melanieclarke commented on jwst/model_blender/blender.py:
```diff
@@ -1,254 +1,152 @@
+            # update ignores list for items in schema that can't be blended
+            self._blend_ignore_attrs.extend(schema_ignores)
+
+            # capture the entire contents of the first model metadata
+            self._first_header_meta = {
+                attr: v
+                for attr, v in model.to_flat_dict(include_arrays=False).items()
+                if attr.startswith('meta') and not any((attr.startswith(i) for i in self._blend_ignore_attrs))
+            }
+
+            # make "blenders" for the metadata with special rules
+            self._blenders = {
+                attr: make_blender(rule)
+                for attr, rule in attr_to_blend_rules.items()
+                # 'first' is the default behavior so these rules need not be made explicitly
+                if rule != 'first' and not any((attr.startswith(i) for i in self._blend_ignore_attrs))
+            }
+
```
This is a style choice so your call, but I think these would be more readable/maintainable if they were written as for loops instead of comprehensions.

melanieclarke replied:
My brain likes that much better.  Thanks!

melanieclarke reviewed (commented):
This is really nice work.  I like the accumulate/finalize mechanism, and the code is much easier to follow than the previous version.

I just have a few clean up requests, mostly documentation related.

braingram commented on jwst/model_blender/schemautil.py:
```diff
@@ -0,0 +1,60 @@
+from stdatamodels import schema as dm_schema
+
+def parse_schema(schema):
+    """
```
Added in e19adb1ff

braingram commented on jwst/model_blender/tablebuilder.py:
```diff
@@ -0,0 +1,117 @@
+    def __init__(self, attr_to_column):
+        """
+        Construct a `TableBuilder` instance using the mapping
+        provided in ``attr_to_column``.
```
Reformated in e19adb1ff

braingram commented on docs/jwst/model_blender/main.rst:
```diff
@@ -27,86 +35,77 @@ This code works by
+The output product will end up with new metadata attribute values and a new ``hdrtab``
+attribute that will produce a ``HDRTAB`` table extension when the product is saved
+to disk.
 
```
I added `ModelBlender` to the docs in 20db503f8

braingram commented:
Thanks for the suggestions and review @melanieclarke I think I addressed all your comments but please let me know if there are any I missed and if you have any follow-up comments from the changes.

melanieclarke commented on jwst/model_blender/rules.py:
```diff
@@ -0,0 +1,112 @@
+Mapping of rule names to functions.
+
+Used for `make_blender`.
+
```
I know it seems a little redundant, since they're listed above, but can you add the list of not-deprecated rules here?  In the built docs, the deprecated ones are easy to see but the supported ones are not.

braingram replied:
I noticed that. I'm not sure why sphinx rendered those in a weird way. I'd rather not duplicate the code in the docs (so they can automatically stay in sync). Although I'd also like to not fight with sphinx for an unknown amount of time to fix the rendering. If the poor rendering is a "deal breaker" I'll duplicate the entries. Let me know either way.

melanieclarke commented on jwst/model_blender/blender.py:
```diff
@@ -1,254 +1,159 @@
+
+        Parameters
+        ----------
+
```
Remove extra newline
```suggestion
```

melanieclarke commented on jwst/model_blender/blendmeta.py:
```diff
@@ -3,328 +3,36 @@
-    """ Create a list of metadata that should be ignored when blending.
-
-    Parameters
+    Properties
```
```suggestion
    Parameters
```

melanieclarke reviewed (commented):
All looks good, just a couple more tiny documentation notes, from re-proofreading.

melanieclarke commented:
> Are the regtests in a state where the changes introduced by this PR can be okified or are there current failures?

They should be pretty clear today, when you're ready to re-run.

braingram commented:
regtest running for okifying: https://plwishmaster.stsci.edu:8081/job/RT/job/JWST/3030/

melanieclarke commented:
Regtest changes look as expected.  @braingram - Are you okifying or do you want me to?

braingram commented:
Thanks for checking. I okified them.
